### PR TITLE
Fixed 'ImageHandler' and 'DatetimeHandler' compound parser shape.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
## Summary

Field handlers gain a shared `normalise()` helper on `AbstractHandler` that converts any loose input shape a caller naturally produces — a bare scalar, a list of scalars, a single keyed record, a list of records, or a mixed list of scalars and records — into a canonical list of records. Callers never need to know whether a field is single- or multi-column, never need to type column names like `target_id` or `value` for the simple case, and the handler always returns a uniform list of records that Drupal entity storage accepts. `ImageHandler::expand()` and `DatetimeHandler::expand()` now use `normalise()` instead of in-line shape sniffing; the three text handlers (`TextHandler`, `TextLongHandler`, `TextWithSummaryHandler`) were migrated from `implements FieldHandlerInterface` to `extends AbstractHandler` so they get the same helper.

The original bug — `ImageHandler::expand()` reading `$values[0]` as a string when the parser handed it `[['target_id' => 'foo.jpg', ...]]` and triggering `TypeError` — is fixed as a consequence. The branch also fixes the `DrupalExtension smoke` workflow that kept breaking whenever `drupalextension` bumped its `minimum-stability` floor.

Consumers (drupal-extension's `EntityFieldParser`, `DrushDriver`, etc.) can now pass whatever they have without inspecting the field's column structure first. That work happens in the consumer packages.

## Changes

- **`src/Drupal/Driver/Core/Field/AbstractHandler.php`** — added `protected function normalise(mixed $values): array`. Discovers the main property via `$this->fieldInfo->getMainPropertyName()`. Accepts five shapes (bare scalar, list of scalars, single record, list of records, mixed list); throws `InvalidArgumentException` for the ambiguous "positional plus named keys at the top level" shape (e.g. `['/path', 'alt' => 'A']`) rather than silently picking an interpretation. Subclasses override `normalise()` only when they have custom shorthand (e.g. a future `NameHandler` "Family, Given" string).

- **`src/Drupal/Driver/Core/Field/ImageHandler.php`** — `expand()` now calls `$this->normalise($values)` and iterates the canonical records. Always returns a list of records (no flat-single-record return, no shape sniffing).

- **`src/Drupal/Driver/Core/Field/DatetimeHandler.php`** — `expand()` now calls `$this->normalise($values)` and iterates the canonical records. Returns a list of `['value' => …]` records with each value formatted for storage.

- **`src/Drupal/Driver/Core/Field/TextHandler.php`**, **`TextLongHandler.php`**, **`TextWithSummaryHandler.php`** — migrated from `implements FieldHandlerInterface` to `extends AbstractHandler`. `expand()` delegates to `normalise()`. Production wiring is unchanged (`Core::getFieldHandler()` already passes the stub/entity_type/field_name constructor args); test setup updated to follow the `newInstanceWithoutConstructor()` + reflection-injected `fieldInfo` pattern.

- **`tests/Drupal/Tests/Driver/Unit/Core/Field/AbstractHandlerNormaliseTest.php`** — new. Single test method `testNormalise` driven by a 16-row data provider. 12 happy-path rows cover every accepted shape against multiple main-property names (`target_id`, `value`, `uri`); 4 throw rows cover the ambiguous-key cases (numeric-0 + named extras, gappy numeric + named, ordering variants). Each row carries `(input, main_property, expected_output, expected_exception_class, expected_exception_message_fragment)` — the test branches on whether an exception class was supplied. A test-local final subclass `NormaliseTestHandler` exists in the same file solely to provide the abstract surface so reflection can invoke the inherited `normalise()`.

- **`tests/Drupal/Tests/Driver/Unit/Core/Field/ImageHandlerTest.php`** — refactored. Two data-provider tests (`testExpandUploadsFile`, `testExpandReusesManagedFile`) now cover every loose shape a caller can naturally pass: bare scalar path, list of paths, single record, list of records, multi-record, plus the same shape variations for the reuse-managed-file path. `createHandler()` injects a `fieldInfo` stub returning `'target_id'`.

- **`tests/Drupal/Tests/Driver/Unit/Core/Field/DatetimeHandlerTest.php`** — `testExpandPreservesEmptyValuesAsNull` is now data-driven (5 shape variants); `createHandler()` stubs both `getSetting('datetime_type')` and `getMainPropertyName()`.

- **`tests/Drupal/Tests/Driver/Unit/Core/Field/TextHandlerTest.php`**, **`TextLongHandlerTest.php`**, **`TextWithSummaryHandlerTest.php`** — updated to construct the handler via `newInstanceWithoutConstructor()` + reflection-injected `fieldInfo`, matching the pattern used by other handler tests.

- **Kernel tests** (`ImageHandlerKernelTest.php`, `ImageHandlerReuseKernelTest.php`, `DatetimeHandlerKernelTest.php`) — inputs use the canonical record shape, which the new `normalise()` returns unchanged.

- **`.github/workflows/ci.yml`** — `DrupalExtension smoke` job: the path repo for `drupal/drupal-driver` is now declared at version `3.0.99` (stable) instead of `3.0.99-alpha`. The previous alpha-suffixed declaration broke every time `drupalextension` bumped its `minimum-stability` floor (alpha → beta → RC). Declaring stable satisfies any floor up to and including stable, decoupling this workflow from the consumer's stability progression.

## Before / After

```
Caller decoupling — ImageHandler

Before (caller must know 'target_id' even for the trivial case):
  $stub->setValue('field_image', [['target_id' => '/path/foo.jpg']]);

After (caller passes whatever's natural; normalise() handles the rest):
  $stub->setValue('field_image', '/path/foo.jpg');                     // bare path
  $stub->setValue('field_image', ['/path/a.jpg', '/path/b.jpg']);      // list of paths
  $stub->setValue('field_image', ['target_id' => '/path/foo.jpg',      // single record
                                  'alt' => 'A', 'title' => 'B']);
  $stub->setValue('field_image', [['target_id' => '/path/foo.jpg',     // list of records
                                   'alt' => 'A']]);
  // All four call sites end up at the same canonical
  //   [['target_id' => N, 'alt' => …, 'title' => …]]
  // inside ImageHandler::expand() and produce the same stored entity.
```

```
ImageHandler::expand() shape pipeline

Before (sniff $values[0], collapse single-record return to flat assoc):
  expand($mixed_input)
    if (is_array($values[0])) treat $values as records
    else                      wrap $values as one record
    foreach each record       resolve target_id
    return count($expanded) === 1 ? $expanded[0] : $expanded
                                  ^ dual return shape

After (canonical pipeline, single return shape):
  expand($mixed_input)
    $records = $this->normalise($values)              // AbstractHandler helper
    foreach each $record                              // always a list
      resolve target_id, default alt/title to NULL
    return list of records                            // always a list
```

```
DrupalExtension smoke workflow — stability coupling

Before (recurring failure pattern):
  drupalextension main composer.json:  "minimum-stability": "RC"
  ci.yml declares path repo:           "drupal/drupal-driver": "3.0.99-alpha"
                                       (alpha < RC -> Composer refuses install)

  Every time drupalextension bumps its stability floor, ci.yml has to follow:
    alpha -> beta -> RC -> stable

After:
  ci.yml declares path repo:           "drupal/drupal-driver": "3.0.99"
                                       (stable -> satisfies any floor up to stable)
```

## Release Notes

* **Refactor**
  * Standardized input normalization across field handlers (text, image, datetime) to consistently process various input shapes and formats.
  * Enhanced handling of empty values and mixed input formats in field handlers.

* **Tests**
  * Expanded test coverage for field handler input normalization and edge cases.

* **Chores**
  * Updated CI workflow for stable version mappings.

[Review Change Stack](https://app.coderabbit.ai/change-stack/jhedstrom/DrupalDriver/pull/370?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/Drupal/Driver/Core/Field/DatetimeHandler.php [hunk 1]
```diff
@@ -22,12 +22,17 @@ public function expand($values): array {
...
 
-    foreach ($values as $key => $value) {
-      $values[$key] = $this->formatDateValue($value, $site_timezone, $storage_timezone);
+    foreach ($values as $value) {
+      // EntityFieldParser emits compound-mode deltas as '['value' => '...']'
+      // arrays. Sibling DaterangeHandler accepts the same shape; mirror it
+      // here so 'datetime' fields work uniformly with the parser.
+      $date = is_array($value) ? ($value['value'] ?? $value[0] ?? NULL) : $value;
+      $formatted[] = $this->formatDateValue($date, $site_timezone, $storage_timezone);
     }
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/Driver/Core/Field/DatetimeHandler.php]: Declining this one. Every sibling handler in `src/Drupal/Driver/Core/Field/` reindexes via `$arr[]`:

- `FileHandler::expand()` → `$files[] = [...]`
- `ImageHandler::expand()` → `$expanded[] = [...]`
- `DaterangeHandler::expand()` → `$result[] = [...]`
- `LinkHandler::expand()` → `$links[] = $link`
- `NameHandler::expand()` → `$names[] = ...`
- `AddressHandler::expand()` → `$result[] = ...`

`DatetimeHandler` now matches that pattern (the whole point of this change was to mirror `DaterangeHandler`). Preserving incoming delta keys here would make `DatetimeHandler` the odd one out without buying anything observable: `EntityFieldParser` only emits numerically-keyed records (0, 1, 2…) and Drupal's entity field storage re-indexes on save, so the keys are never carried through to the stored entity.

If we ever decide that handlers should preserve delta keys, that's a project-wide change across all eight handlers, not a one-off here.

File: src/Drupal/Driver/Core/Field/ImageHandler.php [hunk 2]
```diff
@@ -15,17 +15,29 @@ class ImageHandler extends FileHandler {
...
    */
   public function expand($values): array {
-    $file_path = $values[0];
+    $records = $this->normalise($values);
+    $expanded = [];
 
-    $file = $this->resolveExistingFile($file_path) ?? $this->uploadAndSave($file_path);
+    foreach ($records as $record) {
+      $file_path = (string) $record['target_id'];
+      $file = $this->resolveExistingFile($file_path) ?? $this->uploadAndSave($file_path);
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/Driver/Core/Field/ImageHandler.php:38]: Fixed in 3ea8fa3. ImageHandler::expand() now throws InvalidArgumentException with a clear message when a record is missing target_id, has target_id as NULL, or has it as empty string. Covered by a data-provider-driven test (testExpandThrowsWhenTargetIdMissing) with four rows. Validation is per-handler rather than in normalise() because requirements differ across field types — datetime/text accept empty values, image/file require the path.

[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/Driver/Core/Field/ImageHandler.php:38]: Follow-up to my previous reply: on Alex's suggestion, the key-presence check moved one layer up. AbstractHandler::normalise() now enforces that every output record has the main property key (19a1a79). The ImageHandler-level check is reduced to NULL/empty-string value (key existence is guaranteed by normalise). Three new data-provider rows in AbstractHandlerNormaliseTest cover the missing-key throw path (single record missing key, record in list missing key, empty record).
